### ISSUE TITLE
What metadata/installability checks should Warehouse check uploads for?

### ISSUE THREAD

**Issue by msabramo** [pypi/warehouse#194]
Recently, a project that I worked on pushed a new version to PyPI and it turned out to be completely broken, because the `setup.py` was referencing a `README.rst` file that was not present in the sdist.

It would be awesome if PyPI could so checking of packages that are uploaded.

To start with, it could create a virtualenv and try pip installing the package and make sure that exits with status 0. Perhaps it could also try easy_install to make sure that works too.

There are more elaborate things that could be done like run tests if they are included (many packages don't bundle their tests though) or try to validate the RST, but I think just pip installing is a very good first start, as it detects packages that are completely broken.

**Comment by brainwane** [comment]
This would be a fantastic feature. I'm only moving this to a future milestone because it is not on the critical path for the immediate goal of switching from the old PyPI to Warehouse.

**Comment by jayfk** [comment]
Is this something people are still interested in? 

If we had a system in place that allows to fully install a package there’s lots of other useful information we can extract from it. Transitive dependencies, metadata, even tests.

**Comment by agronholm** [comment]
Yes, people are still interested in this but it's been deferred until the immediate goals have been met and Warehouse has replaced Cheeseshop as the official PyPI.

**Comment by jayfk** [comment]
Best thing is probably to make this a sandboxed standalone project with some kind of API that warehouse can call.

I’m currently working on a POC for this. If anyone is interested, please let me know.

**Comment by brainwane** [comment]
@jayfk I would love for you to post to [pypa-dev](https://groups.google.com/forum/#!forum/pypa-dev) to give a heads-up that you're doing this, so people can speak up there if they're interested in helping.

**Comment by brainwane** [comment]
@jayfk could you post to pypa-dev or [the distutils-sig list](https://mail.python.org/mailman/listinfo/distutils-sig) about your proof of concept? Thanks!

**Comment by brainwane** [comment]
Jannis [posted to distutils-sig](https://mail.python.org/pipermail/distutils-sig/2018-February/031962.html) -- thanks @jayfk!

**Comment by jayfk** [comment]
Yep! Thanks for pointing that out @brainwane!

**Comment by zenogantner** [comment]
Actually, Perl has [CPANTesters](https://qa.perl.org/cpan-testers/), which lets the community help to do the testing over a lot of different platforms and version combinations.
And they have had it for more than a decade.

This is definitely something where Python can learn from Perl.

**Comment by brainwane** [comment]
This is an issue that came up in [a discussion of improving pip's automated tests last week](https://wiki.python.org/psf/PackagingWG/2020-01-23-pip); pip would find this feature helpful.

**Comment by brainwane** [comment]
Reminder to folks following this: as @ewdurbin and @pradyunsg and I [talked about in a meeting about the pip resolver work](https://wiki.python.org/psf/PackagingWG/2020-01-23-pip), getting this implemented might help us smooth testing of and the effects of the resolver rollout. So if any volunteers could help get this finished and merged, we'd appreciate that! cc @yeraydiazdiaz

**Comment by brainwane** [comment]
@uranusjr @pfmoore @pradyunsg I'd appreciate if you could review this issue, especially https://github.com/pypa/warehouse/issues/194#issuecomment-41196203 and the comments below it, which discuss some upload checks we **could** implement in Warehouse (blocking noncompliant uploads). Which of those checks would be particularly valuable to the pip resolver work? Once we have that list, we can make a checklist so this issue becomes more completable.

**Comment by uranusjr** [comment]
One wild wish I’ve always hoped for is to eliminate dynamic dependency altogether, and ensure all dists of a given version specify exactly the same set of dependencies. One of the most resource-consuming (both for development and runtime) part of dependency resolution is the need to download a package matching the host environment, extract it, and potentially build it to get dependencies. This can cost a tremendous amount of time if you’re on a platform without wheels (e.g. musl), and it would be a vast improvement if we are able to download (say) a wheel for Windows and know its metadata would match if I build it on a random Linux machine. m

**Comment by pfmoore** [comment]
It would be great to have a check that the name and version in the metadata match the filename (sdist or wheel), so that we don't have to abort on mismatches. (We'd still have to check, for local files and other edge cases, but knowing that the filename is reliable would still be useful).

As @uranusjr mentioned, the only really useful checks from the pip resolver POV are ones which would allow us to minimise downloading of distributions. All we use is name, version and dependencies.

So for us:

1. As I said, knowing that the filename gives us the correct name/version is massive. Sufficiently so that we make that assumption already and crash out if it turns out to be wrong.
2. Short of Warehouse growing a "give me the metadata" API, we have to download to get dependencies, so the only savings here for us would be across files (i.e., checks that all sdists and wheels for package FOO version X.Y.Z return the same dependency metadata). If we had that, we could cache dependencies for name/version, or extract them from a wheel rather than the corresponding sdist, saving a build step. But we'd still need to special-case PyPI to enable that, so the cost in complexity may mean it's not worth it.

**Comment by dstufft** [comment]
> As I said, knowing that the filename gives us the correct name/version is massive. Sufficiently so that we make that assumption already and crash out if it turns out to be wrong.

I'd argue this is already something you should be able to assume. If those two things don't line up I'd just call it an error. We've never made any promises that it would work, and if it does currently work I'd call it an implementation detail that just happened to allow it for awhile.

> Short of Warehouse growing a "give me the metadata" API

I'm fairly sure that at some point we will supplant the simple API with a better one that will include the needed information, but that is obviously not yet there. That being said, it's probably useful to start checking/enforcing any useful checks now to get better data in the long run.

**Comment by dstufft** [comment]
> eliminate dynamic dependency altogether, and ensure all dists of a given version specify exactly the same set of dependencies. 

I'm not sure that this is possible, or even desirable. At least when we talked about PEP 517 @njsmith argued pretty strongly that we need some sort of hook that enabled programatic dependencies, and that we simply could not depend on static only metadata for sdists.

**Comment by brainwane** [comment]
At the summit last year @crwilcox [asked](https://github.com/pypa/packaging-problems/issues/264#issue-455391624):

> can we fail when there's missing `author`, `author_email`, `URL`. Currently warnings on setup.

@dstufft said, [above](https://github.com/pypa/warehouse/issues/194#issuecomment-41196203) (in 2014):

> - Author -> Sure
> - Author Email -> No
> - URL -> No, sometimes the PyPI Page _is_ the URL

@dstufft do you agree with your 2014 self that author email and project URL should not be mandatory?

**Comment by brainwane** [comment]
@dstufft @pfmoore @pradyunsg @ewdurbin @di @uranusjr @techalchemy  OK, I reread the above discussion. It sounds like the metadata/installability checks that make sense for Warehouse to mandate are:

- Package name
  - must exist
  - must match filename (sdist or wheel)
- Version
  - must exist
  - must be PEP 440-compliant
  - must match filename (sdist or wheel)
- Description: must exist
- Author name: must exist
- Classifiers for supported Python versions - potentially supplanted by #3889/`Requires-Python`
- Certain additional classifiers (which ones?)
  - License

Does this sound right? Are any of these already mandated?

**Comment by pfmoore** [comment]
Sounds reasonable to me.

I'd argue that there should be *some* means of contacting the author, but I concede that "issue tracker on the project webpage" is entirely reasonable for that. So we move out of the area of what should be checked, and into a new metadata field.

**Comment by di** [comment]
* [`Name` is required and validated](https://github.com/pypa/warehouse/blob/643dac748287662d1ffe0eef7012751eb2fa9d7b/warehouse/forklift/legacy.py#L408-L421)
* [`Version` is required and validated](https://github.com/pypa/warehouse/blob/643dac748287662d1ffe0eef7012751eb2fa9d7b/warehouse/forklift/legacy.py#L422-L432)
* [`Description` is optional](https://github.com/pypa/warehouse/blob/643dac748287662d1ffe0eef7012751eb2fa9d7b/warehouse/forklift/legacy.py#L446-L448) (and IMO should remain so)
* [`Author` is optional](https://github.com/pypa/warehouse/blob/643dac748287662d1ffe0eef7012751eb2fa9d7b/warehouse/forklift/legacy.py#L449-L451) (this field is also not required to contain contact information, so might not make sense to require)
* [`Classifiers` is optional](https://github.com/pypa/warehouse/blob/643dac748287662d1ffe0eef7012751eb2fa9d7b/warehouse/forklift/legacy.py#L473-L476) (and IMO should remain so, `Requies-Python` is probably what we should focus on here)

**Comment by miketheman** [comment]
#Triage
I think it's time to close this issue, and here's my reasoning.

Since the update from Dustin in 2020, warehouse has grown a rather robust set of metadata checks, largely driven by adoption of the robust [`packaging` library](https://pypi.org/project/packaging/)'s [metadata module](https://packaging.pypa.io/en/stable/metadata.html) doing most of the heavy lifting.
There's a number of added validations on top of the library, such as which versions of metadata are acceptable, and disallowing local versions - see more in forklift's metadata module:
https://github.com/pypi/warehouse/blob/d9bc68eab2c3e35cb219e1d8cce6a6f23768ce89/warehouse/forklift/metadata.py#L97

There are other issues open to track distinct behaviors related to metadata, such as #17410, #18129, #8481, #8635 that should be solved independently.

If there are other metadata checks that folks think are necessary and not represented in any other open issues, please open a fresh one with a detailed description and reasoning.
